scanpy/preprocessing/_cell_hashing.py

```diff
@@ -0,0 +1,109 @@
+from typing import Iterable
+
+from anndata import AnnData
+
+from scipy.signal import argrelextrema
+from scipy.stats import gaussian_kde
+import numpy as np
+import itertools
+import pandas as pd
+
+
+def _demultiplex_per_barcode(x):
+    """
+    Approximates method in https://www.nature.com/articles/s41592-019-0433-8
+    MULTI-seq: sample multiplexing for single-cell RNA
+    sequencing using lipid-tagged indices
+    """
+    percentiles = np.percentile(x, q=np.arange(.01, 99.9, step=.1))
+    kde = gaussian_kde(x)
+    dist = kde(percentiles)
+    maxima = argrelextrema(dist, np.greater_equal)[0]
+    min_maxima = np.min(maxima)
+    max_maxima = np.max(maxima)
+    intermediate_thresh = (percentiles[min_maxima] + percentiles[max_maxima]) / 2
+    all_low_maxima = maxima[np.where(percentiles[maxima] < intermediate_thresh)[0]]
+    low_maxima = all_low_maxima[np.argmax(dist[all_low_maxima])]
+    info = []
+    for q in range(1, 100):
+        threshold = np.percentile([percentiles[low_maxima],
+                                   percentiles[max_maxima]], q=q)
+        info.append([x >= threshold, threshold])
+    return info, [percentiles, dist]
+
+
+def demultiplex_hashing(adata: AnnData,
+                        percentile_candidates: Iterable[int] = None,
+                        inplace: bool = True):
+    """
+    Demultiplexing cell hashing data
+
+    Params
+    ------
+    adata
+        The annotated data matrix of shape ``n_obs`` × ``n_vars``. Rows correspond
+        to cells and columns to genes.
+    percentile_candidates
+        Choose exact percentiles to evaluate thresholding at
+
+    inplace
+        Whether to update ``adata`` or return dictionary with normalized copies of
+        ``adata.X`` and ``adata.layers``.
+
+    Returns
+    -------
+    Returns `adata.obs` with columns "ID" and "CLASSIFICATION" which identify
+    which cell hashing barcodes a cell has
+    """
+    barcodes_hits = []
+    number_of_barcodes = adata.X.shape[1]
```

Comment:
I am not sure about the logic here. Normally, the columns contain genes, I assume that in cell-hashing there will be barcodes, but could the data be mixed ? Maybe if you clarify what is needed/expected in the description of the function would make the code easy to read.

Reply:
I think these are the columns here represent hashing barcodes, not genes. @njbernstein I think both RNA UMIs and hashing barcode counts should be stored in the same AnnData object. Hashing barcodes can be stored in `adata.obs` or `adata.obsm` but definitely not in `adata.X`. 

I think it makes more sense to first discuss 1) how to store hashing barcodes and 2) how to demultiplex them either in the PR or in the #351 issue.